src/parser/shared/metrics/apl/range.ts

```diff
@@ -0,0 +1,325 @@
+/**
+ * Range checking code for the APL tools.
+ *
+ * Some notes:
+ *
+ * - The overall philosophy here is to assume that the player is in range
+ *   whenever they can be, and let other code (like ABC, or some melee uptime
+ *   code) deal with analyzing whether they could've been in range more than
+ *   they were. Being out of range is thus *not* treated as a mistake by this
+ *   code.
+ *
+ * - The ability to cast a spell in WoW depends on the spell's range being
+ *   greater than the distance to the edge of the target's hitbox. That is:
+ *   `range >= distance - hitbox`, or `range + hitbox >= distance`.
+ *
+ * - Boss hitboxes vary widely in size. Some are very small (Sylvanas, Anduin),
+ *   and some are ENORMOUS (The Jailer, Ragnaros, basically any Dragon-type
+ *   boss). This module tries to pick a hitbox that is relatively accurate
+ *   quickly, and then uses a small (1.5m) margin of error to try to handle
+ *   small differences in our estimate vs the real value. This *usually* picks
+ *   up the actual hitbox within the first handful of events, at least for melee DPS.
+ *
+ * - Hitbox detection relies heavily on using a small number of manual
+ *   `range` entries in a spec's Abilities module to give known-good ranges. For
+ *   melee, the auto-attack range is included by default but this obviously
+ *   doesn't apply to ranged DPS. Adding a single filler or opener spell should
+ *   be enough for it to pick up the hitbox size very quickly.
+ *
+ * - This could very well become a standalone functional analyzer later. Almost
+ *   all the pieces are here.
+ *
+ * @module
+ */
+import Spell from 'common/SPELLS/Spell';
+import ROLES from 'game/ROLES';
+import SPECS from 'game/SPECS';
+import {
+  AnyEvent,
+  HasAbility,
+  HasLocation,
+  HasSource,
+  HasTarget,
+  LocationEvent,
+  ResourceActor,
+  TargettedEvent,
+} from 'parser/core/Events';
+import { encodeTargetString } from 'parser/shared/modules/Enemies';
+import { PlayerInfo } from './index';
+
+const debug = false;
+
+export const Range = {
+  Melee: 5,
+  FullRanged: 40,
+};
+
+const DEFAULT_MELEE_RANGE = Range.Melee;
+const DEFAULT_RANGED_RANGE = Range.FullRanged;
+
+const DEFAULT_HITBOX_SIZE = 0;
+// how far off we're willing to still consider "in range"
+const FUDGE_RANGE = 1.5;
+
+/**
+ * Determine the "default range" for a spell based on spec & role information.
+ *
+ * This is used so that most spells don't require annotations in Abilities.
+ */
+export const defaultRange = (info: PlayerInfo): number => {
+  const role = info.combatant.spec?.role;
+  const spec = info.combatant.specId;
+  switch (role) {
+    case ROLES.DPS.MELEE:
+    case ROLES.TANK:
+      return DEFAULT_MELEE_RANGE;
+    case ROLES.DPS.RANGED:
+      return DEFAULT_RANGED_RANGE;
+  }
```

Comment:
Is there a way to specify a different default for a whole spec, and make it talent dependent? The Druid class tree has a talent that increases the range of everything by 6 yards. Basically all Guardian and Feral druids want this, so typically they're playing with a melee range of 11 yds.

On a related note, I see what these calculations are doing, but I don't understand where/how the results are consumed to shape the output.

Reply:
> Is there a way to specify a different default for a whole spec, and make it talent dependent? The Druid class tree has a talent that increases the range of everything by 6 yards. Basically all Guardian and Feral druids want this, so typically they're playing with a melee range of 11 yds.

I had thought about putting something on the Abilities module for the default range. Maybe I move this there?

> On a related note, I see what these calculations are doing, but I don't understand where/how the results are consumed to shape the output.

The main point that it is used is the [`isInRange`](https://github.com/WoWAnalyzer/WoWAnalyzer/pull/5094/files#diff-fd5c9ce915391510495d8ef8437a8555029c44bc880a4bbbafc9e24f44727e86R263) call here. The init/update state calls are [here](https://github.com/WoWAnalyzer/WoWAnalyzer/pull/5094/files#diff-fd5c9ce915391510495d8ef8437a8555029c44bc880a4bbbafc9e24f44727e86R457) and [here](https://github.com/WoWAnalyzer/WoWAnalyzer/pull/5094/files#diff-fd5c9ce915391510495d8ef8437a8555029c44bc880a4bbbafc9e24f44727e86R381)